pkg/daemon/server/service/rater/rater.go

```diff
@@ -0,0 +1,245 @@
+type Rater struct {
+	pipeline   *v1alpha1.Pipeline
+	httpClient metricsHttpClient
+	log        *zap.SugaredLogger
+	podTracker *PodTracker
+	// timestampedPodCounts is a map between vertex name and a queue of timestamped counts for that vertex
+	timestampedPodCounts map[string]*sharedqueue.OverflowQueue[*TimestampedCounts]
+	options              *options
+}
+
+func NewRater(ctx context.Context, p *v1alpha1.Pipeline, opts ...Option) *Rater {
+	rater := Rater{
+		pipeline: p,
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+			Timeout: time.Second * 1,
+		},
+		log:                  logging.FromContext(ctx).Named("Rater"),
+		timestampedPodCounts: make(map[string]*sharedqueue.OverflowQueue[*TimestampedCounts]),
+		options:              defaultOptions(),
+	}
+
+	rater.podTracker = NewPodTracker(ctx, p)
+	for _, v := range p.Spec.Vertices {
+		// maintain the total counts of the last 30 minutes since we support 1m, 5m, 15m lookback seconds.
+		rater.timestampedPodCounts[v.Name] = sharedqueue.New[*TimestampedCounts](1800)
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			opt(rater.options)
+		}
+	}
+	return &rater
+}
+
+// Function monitor() defines each of the worker's job.
+// It waits for keys in the channel, and starts a monitoring job
+func (r *Rater) monitor(ctx context.Context, id int, keyCh <-chan string) {
+	r.log.Infof("Started monitoring worker %v", id)
+	for {
+		select {
+		case <-ctx.Done():
+			r.log.Infof("Stopped monitoring worker %v", id)
+			return
+		case key := <-keyCh:
+			if err := r.monitorOnePod(ctx, key, id); err != nil {
+				r.log.Errorw("Failed to monitor a pod", zap.String("pod", key), zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *Rater) monitorOnePod(ctx context.Context, key string, worker int) error {
+	log := logging.FromContext(ctx).With("worker", fmt.Sprint(worker)).With("podKey", key)
+	log.Debugf("Working on key: %s", key)
+	podInfo := strings.Split(key, "*")
+	if len(podInfo) != 4 {
+		return fmt.Errorf("invalid key %q", key)
+	}
+	vertexName := podInfo[1]
+	vertexType := podInfo[3]
+	podName := podInfo[0] + "-" + podInfo[1] + "-" + podInfo[2]
+	var count float64
+	activePods := r.podTracker.GetActivePods()
+	if activePods.Contains(key) {
+		count = r.getTotalCount(vertexName, vertexType, podName)
+		if count == CountNotAvailable {
+			log.Debugf("Pod %s does not exist, removing it from the active pod list...", podName)
+			activePods.Remove(key)
+		}
+	} else {
+		log.Debugf("Pod %s does not exist, updating it with CountNotAvailable...", podName)
+		count = CountNotAvailable
+	}
+	// track the total counts using a 10-second time window
+	// e.g. if the current time is 12:00:07, the count will be tracked in the 12:00:00-12:00:10 time window using 12:00:10 as the timestamp
+	now := time.Now().Add(time.Second * 10).Truncate(time.Second * 10).Unix()
```

Comment:
I see a lot of `time.Second * 10`.. can we make it a const so it is easy to update if need be?